### ISSUE TITLE
Renaming private declarations takes 40+ seconds

### ISSUE THREAD

**Issue by FMorschel** [dart-lang/sdk#57031]
While working at https://dart-review.googlesource.com/c/sdk/+/392860 for https://github.com/dart-lang/sdk/issues/49294, I was refactoring and needed to rename a private method for `AddConst`. 

The method was used only once and since it is a file with no parts the analysis and renaming for this should be really fast.

![image](https://github.com/user-attachments/assets/5d91ad16-5571-4026-b30f-669ddd635281)
![image](https://github.com/user-attachments/assets/024a13d4-1a24-481c-ac27-fd1ac1be8291)

After talking with @DanTup on Discord he mentioned that this is actually caused by the creation of an index that for future renamings would decrease the waiting time. 

I was wondering if we could ever do some form of work to help users with this long waiting time for renaming simple things. I have two suggestions which basically overlap:

1. Creating that index in the background hopefully before it is needed in a way that doesn't slow down things the user _does_ care about (suggested by Dan).
2. Basically the above but with a small tweak. I think this could be started whenever the user did request to rename something but we could actually try and leave that work running separately after we had solved the single rename request. I'm not sure how exactly we could do this but I thought of something like:

- User requests to rename something
- We calculate every reference to it that needs to be renamed
- Return that result but also trigger the background process to calculate the index for everything else

**Comment by FMorschel** [comment]
Just suggesting this because in my case I could've simply used find/replace and that would have been way faster. And people who experience this kind of long waiting might use this instead because of this even though there may be some downsides like needing to take a look at Strings and similar things for correct handling of it.

**Comment by bwilkerson** [comment]
@scheglov for verification

> Creating that index in the background ...

I think we already do. Last I looked at it, the index is created (and updated) every time we analyze a library cycle, which means that after the initial analysis of everything that's in the workspace we should have everything we need.

**Comment by scheglov** [comment]
Yes, we compute index when we resolve a Dart file.

We do search across all libraries of a Pub package, including dependencies. So, we need to index all of them. Even if rename never updates dependencies, the search engine, and the index are used for both search and rename. Because rename is search essentially. So, when you do search or rename the first time, we discover and index all dependencies.

**Comment by FMorschel** [comment]
What is happening here then? There is no subprocess or something to help us see inner timings so I'm unsure of what could be happening.

![image](https://github.com/user-attachments/assets/238fc08d-5180-4207-a418-5bdd989d6a6e)

**Comment by scheglov** [comment]
Probably. Yes, we did not have need to break this down into smaller performance operations.

**Comment by DanTup** [comment]
@FMorschel can you reliably reproduce this? I just tried to repro to see if I could get some approx timings with the CPU profiler to confirm the time is where we think (I wondered if some work could be skipped when we know we're only searching for privates?), so I tried:

- Adding a private member to `class AddConst` like `bool _foo()  => false;`
- Restarted VS Code
- Waited for initial analysis to complete (from your screenshot, I don't think you did this during initial analysis)
- Renamed the method

It wasn't fast, but it was only 2-3s which is very different to what you see. That doesn't feel like the kind of difference that differences between our hardware so I think I might not be reproducing the same thing.

**Comment by FMorschel** [comment]
I can. I'm unsure about the real waiting time, but last night, I was working on an assist and renamed the test class (it is just used locally, but it is public), which also took a long time. I'm unsure of what I can do to help identify the problem here. If you have any ideas I'm open to them.

**Comment by DanTup** [comment]
@FMorschel sorry I should've been clearer what I was asking... Can you describe the _precise_ steps so I can see if I can repro in the same way here? Can you also confirm the SDK version you're using for analysis (eg. the one VS Code is selecting... probably your value from `dart.sdkPath`), and whether you're running the server from source (`dart.analyzerPath`) or from the snapshot inside that SDK. Thanks!

**Comment by FMorschel** [comment]
> Can you also confirm the SDK version you're using for analysis

Dart 3.7.0-edge.3a895e28b44aacd08bd9055f05c16f14078eb067 (main) (Wed Nov 6 10:23:38 2024 +0000) on "windows_x64"

> and whether you're running the server from source or from the snapshot inside that SDK

The snapshot =D

---

- Open the SDK project.
- Open `pkg\analysis_server\test\src\services\correction\fix\add_missing_parameter_required_test.dart`
- Rename `AddMissingParameterRequiredTest` to `AAddMissingParameterRequiredTest`
 
![image](https://github.com/user-attachments/assets/3e038844-3954-4fe8-91c5-754d914ecd3c)
![image](https://github.com/user-attachments/assets/d647364f-ac8f-417c-9405-6880ad346264)

About 20s in, I do see that the IDE shows "Analyzing..." on the status bar so some time for that can be reanalyzing files but still. If you now undo this and rename I don't get more than 85ms to rename the same class or `AddMissingParameterRequiredTest_Workspacee` (same file).

If I leave the project open for some time and try to rename it again, it can take ~3s as you mentioned so that is consistent.

**Comment by FMorschel** [comment]
Also, if you think that solving for private declarations is easier you can take a look at `AddConst._declarationListIsFullyConst` in `pkg/analysis_server/lib/src/services/correction/dart/add_const.dart` and rename it.

**Comment by FMorschel** [comment]
I was working on a project at my company with that project and some shared packages we have internally open as well on the IDE. I was definitely not renaming a private declaration, but my first rename took 1 minute and 1 second to complete. I understand this can take a long time to compute for the first rename now, but I think we could maybe start thinking a bit about how to make this a bit faster.

I wonder on larger projects how much time it takes for the first rename to complete. Maybe https://github.com/dart-lang/sdk/issues/60258 can help with this, not sure.

**Comment by DanTup** [comment]
> So, when you do search or rename the first time, we discover and index all dependencies.

@bwilkerson I thought there was a more specific issue about this, but this is the closest issue/comment I could find.

Calling `discoverAvailableFiles` pre-emptively after startup could avoid some of these delays (which can feel random to users that don't know about this), but it's also possible that right after startup the user is trying to do other things that this would then block (IIRC this method always runs completely synchronously).

So I wonder if we should try to make `discoverAvailableFiles` able to run asynchronously (or in batches) so that other work can interrupt it.

**Comment by bwilkerson** [comment]
Possibly. It's been so long since I've looked at that code that I don't have the context to know how easy or hard that would be.

If nothing else, it would be interesting to capture the timing of that operation (maybe initially on the timings page) so that we can determine whether that's really the bottleneck that we think it might be.

@keertip

**Comment by keertip** [comment]
Good point, will take a look at it and see if we can get it on the timings page.
